SEP Proposal: A non-fungible token issuance and metadata specification (#1111)
stellar/stellar-protocol: ecosystem/sep-0040.md

## Preamble

```
SEP: To Be Assigned
Title: Non-Fungible Token Standard
Author: George Kudrayvtsev <george@stellar.org>
Track: Standard
Status: Draft
Created: 2021-01-31
Updated: 2021-01-31
Version: 0.1.0
Discussion: TBD
```

## Simple Summary

> This SEP defines a standard interface to describe non-fungible tokens (NFTs) on the Stellar network, including a distribution mechanism and metadata fields. It draws inspiration from the [EIP-721 standard](https://eips.ethereum.org/EIPS/eip-721).

Reviewer: Suggest using the same terms here as you do later on, so that they are clearly relatable as the same concepts.
```suggestion
This SEP defines a standard interface for non-fungible tokens (NFTs) on the Stellar network, including a issuance model and metadata fields. It draws inspiration from the [EIP-721 standard](https://eips.ethereum.org/EIPS/eip-721).
```

## Motivation

> A standard interface for NFTs allows ecosystem applications (such as wallets or marketplaces) to create, view, or otherwise work with any NFT on Stellar.

Reviewer: This section could be more explicit in that it is motivated by interoperability, for an NFT created in one application, to be viewable, tradable, useable, anywhere Stellar is used.
```suggestion
A standard interface for NFTs that makes NFTs issued by any ecosystem application (such as a wallet or marketplace) accessible and interoperable with any other ecosystem application supporting NFTs.
```

## Abstract

> Non-fungible tokens (NFTs) can describe many types of assets, from digital collectibles to physical items. We should also define a standard issuance model to explain how NFTs can and should be issued on the network. In order to maintain compatibility across the Stellar ecosystem and beyond, we need a standard way to describe and define them that is agnostic of what the NFT actually describes.

Reviewer: The `We should define` is somewhat confusing, should we define it, or is it defined by the protocol? It looks like the protocol defines it, so I think we can say it defines it. Suggested rewording:
```suggestion
Non-fungible tokens (NFTs) can describe many types of assets, from digital collectibles to physical items. This protocol defines a standard set of steps that NFT issuers should use to issue NFTs on the Stellar network, defines metadata that describes an NFT, and specifies how the metadata is linked to the NFT asset on Stellar.
```

Author: I think the `should` there was an editing mishap. I do like your reword, though: it's a good thesis sentence.

## Specification

> We begin with the [issuance model](#issuance-model) and then cover the NFT [metadata specification](#nft-metadata-specification).

Reviewer: It can be helpful for readers if the protocol describes the protocol and doesn't need to tell a story about how the protocol comes to be, so we probably don't need the `We begin` prose. @tomerweller raised a similar comment to this on another protocol we wrote recently with some reasoning that is maybe more elegant than I'm suggesting, so if he chimes in here he might have some thing more thoughtful to say about it. For example:
```suggestion
The protocol defines an [issuance model](#issuance-model) that defines how NFT assets are issued on Stellar, and a [metadata specification](#nft-metadata-specification) that defines how data about the NFT is captured in a way that Stellar guarantees is immutable.
```

I also think that this specification introduction is maybe superfluous and you could drop it if the abstract say substantially the same thing.

Author: Yep, I think this is a good point, I just wanted it to flow a little more naturally (be "prose-y").

### Issuance Model

> Given an arbitrary digital asset, here's how issuance typically works:

Reviewer: ```suggestion
Issuance of NFTs should take the following steps:
```

  1. An account is created specifically to represent the NFT in the Stellar ledger.
  2. The account configures itself with metadata describing the NFT.
  3. The account issues a new Stellar asset representing ownership over said NFT.

>   4. Any account holding that asset lays claim to the NFT.

Reviewer: These steps feel too high-level for the specification. These steps could go into the abstract possibly, or you could keep it here, but if the protocol is specifying how to issue the asset, I think it should do so explicitly, by stating what transactions/operations need submitting, and what values to use, etc.

Reviewer: Or what final state of expected if it's unimportant which operations are used when.

Author: Good idea! I've expanded the issuance model (incl. requiring a distribution account) in 37df0a5; wdyt?

Note also that this SEP does not address concerns around immutability. While there is an optional checksum component (addressed later), the account itself is free to modify the data at-will if it is not locked.

### NFT Specification
This specification is partially inspired by Ethereum's [EIP-721 specification](https://eips.ethereum.org/EIPS/eip-721#specification) on NFTs; however, it is generalized to arbitrary asset types (i.e. not necessarily images).

In order to minimize the costs of on-chain storage, we have multiple levels of indirection of NFT metadata. The only thing the NFT account stores on-chain is a description of where to find the metadata.

#### NFT Account Description
A proper NFT account should specify its `homedomain`. It should also configure the asset in its domain's `stellar.toml` file (see [SEP-1](https://stellar.org/sep-1#currency-documentation)) if appropriate.

There's only one required field on the account itself:

| Field       | Purpose |
| ----------- | ------- |
| `type`      | Describes how the "first hop" of metadata is defined, standard values include `ipfshash` and `stellar` |

> The two possible values for the `type` field describe two standard ways to store the metadata, and the details will be in a data entry corresponding to this value. In other words, `type: ipfshash` means that you should look under the `ipfshash` key data entry which further defines how to find the metadata.

Reviewer 2: What about regular non-ipfs urls? these are pretty common in the eth nft land.  This is important if people will ever try to build nft bridges between eth and stellar.

Author: While I hate the idea of supporting storing assets on a centralized web server, you're right that we should probably include that option here. In fact, both this and the `ipfshash` field could be combined under a single `uri` data entry, but that'd be incompatible with Litemint's current model. I'll add a possible `type: uri` value.

**Backward Compatibility:** Existing NFT platforms essentially default to `type: ipfshash`. While this standard aims to free up the ecosystem from dependence on IPFS, it should also work with existing conventions that developed organically. To do so, **if you see an account with a lone `ipfshash` data entry, you may treat it as an NFT account with `type: ipfshash`**.

##### IPFS Storage
If the `type` is `ipfshash`, the `ipfshash` key must point to an [IPFS Content IDentifier](https://docs.ipfs.io/concepts/content-addressing/) corresponding to the location of the metadata.

If you are the issuer, make sure you understand the implications of using IPFS. You must use an IPFS lookup service to actually resolve to the metadata. For example, [this URL](https://cloudflare-ipfs.com/ipfs/QmaydGEshbL7HMQkD6FZPcMeEt6o15aC4p4P82VxQ7amjy) resolves to the Meridian 2021 Hat NFT via the Cloudflare IPFS service. If you're the issuer, you must also ensure that IPFS actually stores your data (by running your own node or entrusting a provider).

> ##### Stellar Storage

Reviewer 2: Can we remove stellar storage altogether? Storing binary data on-chain should not be encouraged. Maybe we can add an FAQ section explaining why it's not included?

Author: There was a key motivation to adding this section: adding more than one `type`. With the (pending) addition of `uri` in response to your [comment below](https://github.com/stellar/stellar-protocol/pull/1111#discussion_r797231832), this makes sense to drop in light of your response.

I still think keeping things on-chain is way cooler than other NFT solutions, but I understand why it's not something we want to encourage.

> If the `type` is `stellar`, the `stellar` key must point to a Stellar account ID. That account will store the metadata within its data entries. Refer to the unofficial [SEP-39](https://github.com/stellar/stellar-protocol/pull/1090) specification for details on how to read an account's data entries for the NFT metadata.

Reviewer: There isn't anything such as an official or unofficial SEP in my opinion, if that SEP is SEP-39 we should merge it, or if not drop this section until SEP-39 is merged.
```suggestion
If the `type` is `stellar`, the `stellar` key must point to a Stellar account ID. That account will store the metadata within its data entries. Refer to [SEP-39](SEP-0039.md) specification for details on how to read an account's data entries for the NFT metadata.
```

  "url": "https://ipfs.io/ipfs/QmTAQ1e5zD4rxEspDgSZNgqXb73LBEj2wpQkDTj1UvCmv6",
  "code": "MERIDIAN2021",
  "issuer": "GDOA2FECYKBVG2RQYG5JQO6EBLCDBWZ7MFBEZY6AIBJ4NDXV3XEVCVBC",
  "domain": "meridian.litemint.store",
  "supply": 50,
  "video": "ipfs://QmeDxfvNpoizLxP4xLNrx1QuykPSqXEyZL8tVceZGudcvs"
}
```

We can see that this account [really does](https://horizon.stellar.org/operations/164563600389443587) own and issue this asset, meaning the NFT is genuine.

## Design Rationale
TODO.

## Security Concerns
While there are no direct security concerns to the network introduced by this SEP, the trust model of NFTs deserves some clarification.

Digital assets are infinitely and perfectly reproduceable by virtue of being digital. Thus, the goal of an NFT should **not** be to represent ownership or possession of a particular digital asset, but rather to represent the existence of a particular _relationship_ with the issuer.

Those processing an NFT should take care to ensure that the `issuer` field matches the account referencing the NFT, and that the `code` is in fact an asset issued by said account. Without these two links, you cannot make any conclusions about the NFT.

> There is another trust issue worth nothing: **supply**. This metadata field _should not be trusted_, and should instead be cross-referenced with the account's distribution operations. Unless an account is completely locked, nothing inherently prevents it from ignoring this `supply` field and issuing new tokens.

Reviewer: Why include `supply` in the metadata if it requires verification from another source? Is it intended to capture the supply at the time of mint?

Author: It's included for compatibility with Litemint, see [here](https://cloudflare-ipfs.com/ipfs/QmaydGEshbL7HMQkD6FZPcMeEt6o15aC4p4P82VxQ7amjy). However, I was hesitant to add it, so I'm fine leaving it off of the official spec.

Reviewer: It'd simplify the spec to leave it out. Any field included should be very well defined and intentional imo. Services like litemint can always define extra fields or we can add it later.

@FredericRezeau what is supply? Is it required by litemint?

Author: I dropped references to `supply` in 87d6a3bb8367d7ff5285772cd9a6f381ff44dc44, but will leave this open if Frederic wants to chime in.

  3. The account issues a new Stellar asset representing ownership over said NFT.
  4. Any account holding that asset lays claim to the NFT.

The NFT model involves two levels of indirection in order to minimize on-chain storage:

    +---------------+      +----------+      +---------+
    |  Stellar NFT  | ---> | metadata | ---> |   NFT   |
    |  account data |      |   JSON   |      | content |
    +---------------+      +----------+      +---------+

## Specification
This SEP defines an [issuance model](#issuance-model) that defines how NFT assets are issued on Stellar and a [metadata specification](#nft-metadata-specification) that defines how data about the NFT is captured. The metadata is stored in a way such that the Stellar network guarantee its authenticity.

### Issuance Model
Standard issuance of NFTs must take the following steps:

  1. Create an account to represent the NFT: `CreateAccountOp(A, <reserve>)`
  2. Configure the NFT's metadata pointer, as described [below](#nft-account-description): `ManageDataOp("type", "...")`
  3. Form an asset describing your NFT: `Code:Issuer` (where issuer is `A`)

>   4. Create a distribution account: `B = CreateAccountOp()` and `ChangeTrust("Code:Issuer", <limit>)`

Author: Another great idea, done.

>   5. Issue the tokens: `PaymentOp(B, "Code:Issuer", <amount>)`

Author: I amended it to put the pseudocode in a separate block, wdyt?

  6. (Optional) Lock the issuing account: `SetOptionsOp(MasterWeight: 0)`

> Note that without this final step, there are concerns around both immutability and supply: an account may modify its NFT metadata pointer or issue more tokens unless it's locked down. Thus, Step 6 is recommended.

Author: Great point re: impact; addressed.

Ownership of the token directly conveys ownership of a unit of the NFT, however that may be defined. Other scenarios such as ownership transfers via payments are built on top of this basic issuance model and are thus extraneous to this SEP.

Other issuance models may exist and may be compatible with the following specification, but this is the standard issuance model on Stellar.

### NFT Specification
This specification is partially inspired by Ethereum's [EIP-721 specification](https://eips.ethereum.org/EIPS/eip-721#specification) on NFTs; however, it is generalized to arbitrary asset types (i.e. not necessarily images).

In order to minimize the costs of on-chain storage, we have multiple levels of indirection of NFT metadata. The only thing the NFT account stores on-chain is a description of where to find the metadata.

#### NFT Account Description

> A proper NFT account should specify its `homedomain`. It should also configure the asset in its domain's `stellar.toml` file (see [SEP-1](https://stellar.org/sep-1#currency-documentation)) if appropriate.

Author: There are two parts: we already have a way to make an asset "verified" (via the toml) and also "render" it in wallets (logos and such). NFTs should be compatible with those existing methods, but also have their own spec for _new_ tools.

There are two **required** fields on the account itself:

| Field       | Purpose |
| ----------- | ------- |

> | `type`      | Describes how the "first hop" of metadata is defined, standard values include `ipfshash` and `uri` |

Author: Good point. Since `ipfshash` is a backwards-compatibility thing, `url` is a catch-all that should future-proof things.

> | `type`      | Describes how the "first hop" of metadata is defined, standard values include `ipfshash` and `uri` |

Reviewer 3: We should really discourage using many subentries here to avoid further ledger bloat

> | `type`      | Describes how the "first hop" of metadata is defined, standard values include `ipfshash` and `uri` |

Author: Agreed, we're trying to heavily limit them and keep most data off-chain.

> | `hash`      | The [SHA3](https://csrc.nist.gov/publications/detail/fips/202/final) hash of the metadata contents |

Reviewer 4: I came here to chime in with the same thought. SHA3-256 is not an "upgraded" version of SHA2-256, it is considered an alternative should a break in the SHA2 family be found.

I would make the suggestion that perhaps there should be a 'hashType' field in addition to the 'hash' field. This would allow any arbitrary hash function type to be used (e.g. `sha`, `sha2-256`, `sha2-512`, `sha3-256`, `sha3-512`, `blake3`). If this approach is taken an enum of accepted types could be consulted to determine if what is being submitted is an acceptable type. I would continue to allow SHA1, even though it is not considered to be as secure, only to handle the case where content can be stored in a public Git repository and pointed to by the hash of its commit.

I would also suggest specifying the hash encoding as part of the spec. Hexadecimal encoding is the more popular way to represent hashes, but Base64 is also common and more compact. You should choose one and validate it on submission.

Cheers.

> | `hash`      | The [SHA3](https://csrc.nist.gov/publications/detail/fips/202/final) hash of the metadata contents |

Reviewer: Storing hashType assumes a single hash type and so I think we store different hashes types in unique fields and applications can simply look for their fields for the hash types they support. i.e. store the sha256 in a `sha256` field. If someone decided to store a sha-3 hash too, they can instead or in addition store a `sha3` field.

We don't need to encode the hashes, but should be explicit and clear that the hash is stored in it's binary form in the data value. Applications can render the hash in whatever encoding they find most appropriate.

> | `hash`      | The [SHA3](https://csrc.nist.gov/publications/detail/fips/202/final) hash of the metadata contents |

Reviewer 4: I would take these comments a step further and *require* that a hash of the content be provided in all cases. It has been a *major* shortcoming of existing NFT specs to allow linking to content that is not committed to by a hash of the content. It is so easy to hash the bytes of any content that it should never be skipped.

This can be the case even if IPFS is chosen as the `url`. An IPFS CID is not a file content hash and is harder for people to verify independently.

See:

https://docs.ipfs.io/concepts/hashing/#content-identifiers-are-not-file-hashes

By including a `hash` and `hashType` you make it much easier for client libs to verify the integrity of the file content using commonly found hashing libraries or stdlib functions. The IPFS hash is application specific code and its method of verification may change over time and should be considered a bonus integrity check on top of the content hash.

> | `hash`      | The [SHA3](https://csrc.nist.gov/publications/detail/fips/202/final) hash of the metadata contents |

Author: Sha3 was supposed to (a) future-proof and (b) align with Ethereum, but I appreciate the benefit of sticking to what we already use.

> | `hash`      | The [SHA3](https://csrc.nist.gov/publications/detail/fips/202/final) hash of the metadata contents |

Reviewer 4: Side note. Ethereum uses Kechack256 which is not the same as SHA3-256. Different internal padding results in different hash outputs. So using SHA3 would not end up aligning with Ethereum in any case.

https://ethereum.stackexchange.com/questions/30369/difference-between-keccak256-and-sha3#30373

> | `hash`      | The [SHA3](https://csrc.nist.gov/publications/detail/fips/202/final) hash of the metadata contents |

Author: Blegh, I hate that GitHub doesn't show me new comments as they come in. My comment was only in reply to leigh's first comment...

> | `hash`      | The [SHA3](https://csrc.nist.gov/publications/detail/fips/202/final) hash of the metadata contents |

Author: Good thoughts here, thanks both! So we'll have an enumeration of possible keys that correlate to hashes? I don't really like that, and would prefer a type field. It makes it easier to look it up, no? Otherwise you have clients doing strange for-loops looking for a hash entry.

Regarding IPFS, this may be a product of my own ignorance, but I think CIDs still give you integrity, right? You know that the content at the CID is exactly the content put there by the issuer since if it changed, the CID would also change. There's no _need_ to verify that hashes match, since it's baked into the CID, right? To be clear, I do still think the hash is valuable for integrity (and especially in the non-IPFS case), I'm just trying to clarify my own understanding.

> ##### `type`

Reviewer: Actually, the big advantage to not using something like an nft namespace is the interoperability with tokens that maybe aren't conceptually NFTs but are otherwise compatible. This could also be seen as a disadvantage, but I think it's an advantage.

The two possible values for the `type` field describe two standard ways to store the metadata, and the details will be in a data entry corresponding to this value. In other words, `type: ipfshash` means that you should look under the `ipfshash` key data entry which further defines how to find the metadata.

All other values for the `type` are considered non-standard until they are approved for inclusion within this SEP.

##### `hash`
The `hash` is provided to ensure integrity of the metadata that the account is pointing to. This prevents the data at the destination from being modified without explicit knowledge and approval by the issuer.

> If it is excluded, the NFT **must not be considered genuine**, _unless_ the `type` is `ipfshash`, since the IPFS CID includes an integrity check on the data it points to. 

Author: If the hash doesn't match then you have absolutely have no guarantees about the authenticity of the metadata. I'll reword this :+1:

##### Storage
Fields should be stored as simple key-value pairs within an account's data entries. The standard limitations of `ManageDataOp` apply: field names and their values can only be 64 bytes each. Most metadata attributes should be stored at the NFT reference location itself.

**Backward Compatibility:** Existing NFT platforms essentially default to `type: ipfshash`. While this standard aims to free up the ecosystem from dependence on IPFS, it should also work with existing conventions that developed organically. To do so, **if you see an account with a lone `ipfshash` data entry, you may treat it as equivalent to an NFT account with `type: ipfshash`**.

##### IPFS Storage
If the `type` is `ipfshash`, the `ipfshash` key must point to an [IPFS Content IDentifier](https://docs.ipfs.io/concepts/content-addressing/) corresponding to the location of the metadata.

> If you are the issuer, make sure you understand the implications of using IPFS. You must use an IPFS lookup service to actually resolve to the metadata. For example, [this URL](https://cloudflare-ipfs.com/ipfs/QmaydGEshbL7HMQkD6FZPcMeEt6o15aC4p4P82VxQ7amjy) resolves to the Meridian 2021 Hat NFT via the Cloudflare IPFS service. If you're the issuer, you must also ensure that IPFS actually stores your data (by running your own node or entrusting a provider).

Author: Really nice clarification, thanks!

##### URL Storage
If the `type` is `uri`, the corresponding `uri` key must point to a [valid URI](https://datatracker.ietf.org/doc/html/rfc3986) containing the NFT's metadata. 

_Note_: While you could technically describe metadata living on the IPFS with a URL by pointing to a specific node or a URI prefixing the CID with `ipfs://`, we recommend describing _only_ its CID to leave lookups to the users.

#### Standard Metadata Specification
This section describes the metadata itself rather than how to find it. 

The metadata must be valid JSON and must contain the following **required** fields:

| Field       | Description |
| ----------- | ----------- |
| `name` | a short name or title for the NFT |
| `description` | a longer, more-detailed description of the NFT |
| `type` | a [Media Type](https://www.iana.org/assignments/media-types/media-types.xhtml) describing the NFT's primary format (pointed to by `url`) |

> | `uri` **or** `url`  | a [URI](https://datatracker.ietf.org/doc/html/rfc3986) pointing to the actual NFT contents |

Reviewer: Could we pick one field name, or do we need both for consistency with other standards?

Author: Well I want it to be `uri`, since that's broader, but Litemint already uses `url`, so I shoe-horned both in here. I'm not happy about it, though. We could say that the name is `url` but it can be any URI??

Reviewer: A URI is just an identifier, a URL is a locator. Given that the intention here is to define a location that the NFT is stored, I think URL is fine and easily understood by a larger audience anyway.

There is probably not a huge amount of utility in saying it can be an identifier that isn't a locator, so I don't think URI is worth it.

Reviewer: To be clear `ipfs://...` is a URL.

Reviewer 4: URL. This is the way.

https://danielmiessler.com/study/difference-between-uri-url/

Author: That's a good point, I think I mixed myself up double-checking my statements against too many RFCs.

| `issuer` | a Stellar public key ([`G...`](https://stellar.org/protocol/sep-23)) corresponding to the account which issued this NFT |
| `code` | the NFT's asset code |

We allow either `uri` or `url` to maintain compatibility with existing ecosystem conventions. If _both_ are present, prefer the `uri`.

For example,

```json
{
  "name": "A demonstration of NFT metadata",
  "description": "This is a description of the cool NFT used in this demo.",
  "type": "image/png",
  "url": "https://cloudflare-ipfs.com/ipfs/QmXoypizjW3WknFiJnKLwHCnL72vedxjQkDDP1mXWo6uco/I/m/RickAstleyNeverGonnaGiveYouUp7InchSingleCover.jpg",
  "issuer": "GALAXYVOIDAOPZTDLHILAJQKCVVFMD4IKLXLSZV5YHO7VY74IWZILUTO",
  "code": "DEMOASSET"
}
```

> 

Reviewer 4: Why not support the concept of a single NFT pointing to multiple pieces of content as an array of content items? This would allow a single NFT to point to a metadata file that in turn has pointers to multiple resources (e.g. a video, a thumbnail of the video, a certificate document).

This could be represented by a `data` top-level property that always contains an Array of at least one object. each object in the array could have additional attributes.

For example,

```json
{
  "name": "A demonstration of NFT metadata",
  "description": "This is a description of the cool NFT used in this demo.",
  "data": [
    {
     "description": " This is a cool image",
     "type": "image/png",
     "url": "https://filename.jpg",
     "hash": "deadbeef01",
     "hashType": "sha2-256",
    },
    {
     "description": " This is a cool video",
     "type": "video/mpeg",
     "url": "https://filename.mp3",
     "hash": "deadbeef02",
     "hashType": "sha2-512",
    }
  ],
  "issuer": "GALAXYVOIDAOPZTDLHILAJQKCVVFMD4IKLXLSZV5YHO7VY74IWZILUTO",
  "code": "DEMOASSET"
}
```

This would allow new types of multi-part assets to be referenced in a single NFT.

Something that will have to be addressed in any case is also how to canonicalize the contents of the metadata file so that it can be hashed consistently by both the creator and the verifier. JSON objects are not guaranteed to have their contents in any particular order. A standard method for ordering the metadata file so that it can be consistently hashed by all parties must be defined. Otherwise, it may not be possible to verify the on-chain metadata content hash.

See for example:

https://github.com/SkeLLLa/node-object-hash#readme

Author: The multi-part suggestion is really nice, thanks! We want to maximize compatibility, though, so we may have to resort to "either a `data` array or the same fields on the root object" which I hate but aligns with existing NFTs.

Regarding JSON ordering: The hash should be computed on the raw metadata file rather than the "loaded" JSON blob in order to avoid this. It's an important subtlety, so I'll note it.

Reviewer: We could support the URL/ipfshash being to a directory, and inside the directory each asset code could get its own .json, so that an account can issue multiple asset codes but share the issuing account. This should work with most URL protocols.

#### Standard Metadata Specification
This section describes the metadata itself rather than how to find it.

The metadata must be valid JSON and must contain the following **required** fields:

| Field       | Description |
| ----------- | ----------- |
| `name` | a short name or title for the NFT |
| `description` | a longer, more-detailed description of the NFT |
| `type` | a [Media Type][mediatype] describing the NFT's primary format (pointed to by `url`) |
| `url`  | a [URL][url] pointing to the actual NFT contents |
| `issuer` | a Stellar public key ([`G...`](https://stellar.org/protocol/sep-23)) corresponding to the account which issued this NFT |
| `code` | the NFT's asset code |

For example,

```json
{
  "name": "A demonstration of NFT metadata",
  "description": "This is a description of the cool NFT used in this demo.",
  "type": "image/png",
  "url": "https://cloudflare-ipfs.com/ipfs/QmXoypizjW3WknFiJnKLwHCnL72vedxjQkDDP1mXWo6uco/I/m/RickAstleyNeverGonnaGiveYouUp7InchSingleCover.jpg",
  "issuer": "GALAXYVOIDAOPZTDLHILAJQKCVVFMD4IKLXLSZV5YHO7VY74IWZILUTO",
  "code": "DEMOASSET"

> }

Reviewer: Could we add the hash of the asset in the metadata?

Author: Good call, but I'll echo what I said [above](): I still don't think we should have an enumeration-based approach for hashing. We should either have a type (so `type: sha256` and then `sha256: <hash>`), a standard hashing algorithm under `hash:`, or some short whitelist of hashes (like `sha256: <hash>`, `blake3: <hash>`, etc.)

Reviewer 4: > Could we add the hash of the asset in the metadata?

If I understand what you are proposing @leighmcculloch, I think you would lose the major benefit of having a cryptographic commitment to the content on-chain. The hash that commits to the metadata file must be on-chain. A malicious hoster of the metadata file could change the content of the metadata (including an embedded hash) to entirely new content.  This is similar to the problem that exists today with ERC721 were images hosted and linked to from an NFT could be swapped out or removed with the holder of the NFT not being able to detect that change.

> I still don't think we should have an enumeration-based approach for hashing. 

I would concur with this opinion.  I think there is a lot to be said for the simplicity of a `hash` field that contains arbitrary hash bytes, and a `type` or `hashType` field that contains the description of what the `hash` bytes represent. This is the scheme we are taking at Truestamp.

With this scheme there is always a consistent property name to find what you are looking for. It was mentioned that having a JSON key per hash type allows for multiple hashes to be associated with an NFT, but is this a real use-case? Under which circumstances would a developer need to store both a SHA2-256 and BLAKE3 hash of the same metadata file?

This same scheme can then be applied to having a `data` array at the top level of the metadata file that contains an array of objects representing at least one, but potentially many, image, video, document, etc objects that the metadata points to.  The metadata should also then have a `hash` and `hashType` field for each of those content objects that commits to the hash of the contents of each file. The same `hashType` does not have to be used for each content object in the `data` array.

This affords a simplicity for the developer who can always just look for a single `hash` and `hashType` value at both the on-chain and metadata levels to verify that the metadata file, and all content objects it links to, are probably unchanged..

Author: > If I understand what you are proposing

I interpreted this as Leigh wanting an _additional_ hash of the file contents :thinking: Otherwise, I completely agree with you re: integrity on-chain. You could have the issuer sign the metadata and put that in the data itself, but that seems overly convoluted. We just need integrity, since the chain itself provides authenticity.

> I think there is a lot to be said for the simplicity of a hash field that contains arbitrary hash bytes, and a type or hashType field that contains the description of what the hash bytes represent. 

Is there any reason why we wouldn't just stick to a particular hashing algorithm? Every data entry saved is a win in our book.

I really liked the `data` array approach, but after Fred's comments (see below in the issue), it's a little unclear to me where and how we're going to describe NFTs in general now.

Reviewer: I was referring to adding a `sha256` field to the metadata JSON that contained a hash of the asset `url` file, assuming that the url is stored somewhere that isn't intrinsically verifiable like IPFS is.

Reviewer: I agree this discussion is mostly not relevant if we are exploring a focus on SEP-1.
